solutions/platinum/cses-1160.mdx

```diff
@@ -199,4 +199,88 @@ int main() {
 ```
 
 </CPPSection>
+<JavaSection>
+
+```java
+
+import java.io.*;
+import java.util.StringTokenizer;
+
+public class PlanetQueries2 {
+	static Kattio x = new Kattio();
+	static PrintWriter pw = new PrintWriter(System.out);
+	static int n;
+	static int q;
+	static int[] arr;
+	static int[][] binaryLifting;
+	static int[] len;
+
+	public static void main(String[] args) {
+		n = x.nextInt();
+		q = x.nextInt();
+		arr = new int[n + 1];
+		for (int i = 1; i <= n; i++) { arr[i] = x.nextInt(); }
+		binaryLifting = new int[n + 1][20];
+		len = new int[n + 1];
+		/*
+		 * in preprocess, doing dfs, for each node, to ensure to covering
+		 * more than one connected component
+		 */
+		for (int i = 1; i <= n; i++) { binaryLifting[i][0] = arr[i]; }
+		// visited array ensures to cover more than one connected components
+		boolean visited[] = new boolean[n + 1];
+		for (int i = 1; i <= n; i++) {
+			if (!visited[i]) { dfs(i, visited); }
+		}
+		StringBuilder output = new StringBuilder();
+		for (int i = 0; i < q; i++) {
+			int a = x.nextInt();
+			int b = x.nextInt();
+			int aa = jump(a, len[a]);  // in case, a is in front of b
+			if (jump(a, len[a] - len[b]) == b) {
+				pw.println(len[a] - len[b]);
+			} else if (jump(aa, len[aa] - len[b]) == b) {
+				pw.println((len[aa] - len[b]) + len[a]);
+			} else {
+				pw.println(-1);
+			}
+			/*
+			 * in any of the edge cases, for example, (u,v) u being in cycle and
+			 * v is in tree u will never be able to "jump" to v, and this is
+			 * precisely being checked in the above if conditions.
+			 */
+		}
+		pw.close();
+	}
+	static void dfs(int node, boolean visited[]) {
+		if (visited[node]) { return; }
+		visited[node] = true;
+		dfs(arr[node], visited);
+		/*
+		 * head recursion is being used here, in order to start counting the
+		 * height of the graph, from the left nodes of the tree (if one
+		 * exists...)
+		 */
+		len[node] = len[binaryLifting[node][0]] + 1;
+		for (int level = 1; level < 20; level++) {
+			binaryLifting[node][level] =
+			    binaryLifting[binaryLifting[node][level - 1]][level - 1];
+		}
+	}
+	static int jump(int a, int dist) {
+		if (dist < 0) return -1;
+		int level = 0;
+		while (dist != 0) {
+			if ((dist & 1) == 1) { a = binaryLifting[a][level]; }
+			++level;
+			dist = dist >> 1;
```

Comment:
just divide it by 2 instead

Reply:
right-shifting and dividing by 2 are the same thing, right ?

Reply:
more readable

Reply:
cool. updated and checked for AC.